pkg/controller/ippool/defaults.go

```diff
@@ -0,0 +1,330 @@
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ippool
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/netip"
+	"strings"
+
+	configv1 "github.com/openshift/api/config/v1"
+	operator "github.com/tigera/operator/api/v1"
+	crdv1 "github.com/tigera/operator/pkg/apis/crd.projectcalico.org/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// cidrToName returns a valid Kubernetes resource name given a CIDR. Kubernetes names must be valid DNS
+// names. We do the following:
+// - Expand the CIDR so that we get consistent results and remove IPv6 shorthand "::".
+// - Replace any slashes with dashes.
+// - Replace any : with dots.
+func cidrToName(cidr string) (string, error) {
+	// First, canonicalize the CIDR. e.g., 192.168.0.1/24 -> 192.168.0.0/24.
+	_, nw, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return "", err
+	}
+
+	// Parse the CIDR and expand it to its full form.
+	// e.g., fe80::/64 -> fe80:0000:0000:0000:0000:0000:0000:0000/64
+	pre, err := netip.ParsePrefix(nw.String())
+	if err != nil {
+		return "", err
+	}
+	name := pre.Addr().StringExpanded()
+
+	// Replace invalid characters.
+	// e.g., fe80:0000:0000:0000:0000:0000:0000:0000/64 -> fe80.0000.0000.0000.0000.0000.0000.0000-64
+	name = strings.ReplaceAll(name, ":", ".")
+	name += fmt.Sprintf("-%d", pre.Bits())
+
+	return name, nil
+}
+
+// fillDefaults fills in IP pool defaults on the Installation object. Defaulting of fields other than IP pools occurs
+// in pkg/controller/installation/
+func fillDefaults(ctx context.Context, client client.Client, instance *operator.Installation, currentPools *crdv1.IPPoolList) error {
+	if instance.Spec.CNI == nil || instance.Spec.CNI.IPAM == nil {
+		// These fields are needed for IP pool defaulting but defaulted themselves by the core Installation controller, which this controller waits for before
+		// running. We should never hit this branch, but handle it just in case.
+		return fmt.Errorf("Cannot perform IP pool defaulting until CNI configuration is available")
+	}
+
+	// Only add default CIDRs if there are no existing pools in the cluster. If there are existing pools in the cluster,
+	// then we assume that the user has configured them correctly out-of-band and we should not install any others.
+	if currentPools == nil || len(currentPools.Items) == 0 {
```

Comment:
Since this includes `len(currentPools.Items) == 0` is there a way to have the operator wait for pools to be created externally?

Reply:
The way to do that is to create IP pools before creating the Installation, or to create the installation with an empty IP pools list like this:

```yaml
spec:
  calicoNetwork:
    ipPools: []
```